sandbox/grist/formula_prompt.py

```diff
@@ -0,0 +1,145 @@
+  except (ValueError, KeyError):
+    return 'str'
+
+
+def values_type(values):
+  types = set(type(v) for v in values) - {RaisedException}
+  optional = type(None) in types
+  types.discard(type(None))
+
+  if types == {int, float}:
+    types = {float}
+
+  if len(types) != 1:
+    return "Any"
+
+  [typ] = types
+  val = next(v for v in values if isinstance(v, typ))
+
+  if isinstance(val, records.Record):
+    type_name = val._table.table_id
+  elif isinstance(val, records.RecordSet):
+    type_name = "List[{}]".format(val._table.table_id)
+  elif isinstance(val, list):
+    type_name = "List[{}]".format(values_type(val))
+  elif isinstance(val, set):
+    type_name = "Set[{}]".format(values_type(val))
+  elif isinstance(val, tuple):
+    type_name = "Tuple[{}, ...]".format(values_type(val))
+  elif isinstance(val, dict):
+    type_name = "Dict[{}, {}]".format(values_type(val.keys()), values_type(val.values()))
+  else:
+    type_name = typ.__name__
+
+  if optional:
+    type_name = "Optional[{}]".format(type_name)
+
+  return type_name
+
+
+def referenced_tables(engine, table_id):
+  result = set()
+  queue = [table_id]
+  while queue:
+    cur_table_id = queue.pop()
+    if cur_table_id in result:
+      continue
+    result.add(cur_table_id)
+    for col_id, col in visible_columns(engine, cur_table_id):
+      if isinstance(col, BaseReferenceColumn):
+        target_table_id = col._target_table.table_id
+        if not target_table_id.startswith("_"):
+          queue.append(target_table_id)
+  return result - {table_id}
+
+
+def visible_columns(engine, table_id):
+  return [
+    (col_id, col)
+    for col_id, col in engine.tables[table_id].all_columns.items()
+    if is_visible_column(col_id)
+  ]
+
+
+def class_schema(engine, table_id, exclude_col_id=None):
+  result = "@dataclass\nclass {}:\n".format(table_id)
+  for col_id, col in visible_columns(engine, table_id):
+    if col_id != exclude_col_id:
+      result += "    {}: {}\n".format(col_id, column_type(engine, table_id, col_id))
+  result += "\n"
+  return result
+
+
+def get_formula_prompt(engine, table_id, col_id, description):
+  result = ""
+  for other_table_id in sorted(referenced_tables(engine, table_id)):
+    result += class_schema(engine, other_table_id)
+
+  result += class_schema(engine, table_id, col_id)
+
+  return_type = column_type(engine, table_id, col_id)
+  result += "    @property\n"
+  result += "    def {}(self) -> {}:\n".format(col_id, return_type)
+  result += '        """\n'
+  result += '{}\n'.format(textwrap.indent(description, "        "))
+  result += '        """\n'
+  return result
+
+
+def convert_completion(completion):
+  result = textwrap.dedent(completion)
+  replacements = []
+  atok = ASTTokens(result, parse=True)
+  for node in ast.walk(atok.tree):
+    if isinstance(node, ast.Name) and node.id == "self":
+      start, end = atok.get_text_range(node)
+      # Avoid $ because of f-strings
```

Comment:
Hmm I got a weird `recrecreturn f"{self.Date.year}-{self.Date.month}"` result where `return f"{rec.Date.year}-{rec.Date.month}"` was probably what was intended, is there some library version issue or other wildcard to watch out for here @alexmojaki do you happen to know?

Reply:
Yes, that makes sense, I just didn't think of this. Better add a case to the test for `convert_completion` with `self` inside f-strings.

The fix is to replace `ASTTokens(result, parse=True)` with `ASTText(result)` which requires upgrading `asttokens` to the latest version. Doing this will fix the bug you found but still won't allow using `$` instead of `rec`.

Somewhat related is https://phab.getgrist.com/D3690, which does similar upgrades in the main code. And if everything there is done in full, including the astroid stuff, then using `$` here will become safe.

Reply:
Thanks for advice @alexmojaki. I tried with ASTText and ver 2.1.0 but still didn't see fixes happen within fstring. Started looking at asttokens but tempted to just brutally define self = rec for this demo.

Reply:
Oh, also you need to be using Python >= 3.9.7. The patch version matters.

Reply:
Ah, that was it - thanks @alexmojaki